feat: add tree-sitter-make

EricCrosson opened the pull request:
This commit adds syntax highlighting for GNU Make[^1] makefiles
via tree-sitter-make[^2].

[^1]: https://www.gnu.org/software/make/
[^2]: https://github.com/alemuller/tree-sitter-make

EricCrosson commented:
Hey guys! Helix is an _awesome_ project, I've only found it in the last few days but I love what I've seen so far. Thanks so much for your hard work! What a great idea to build an editor from the ground up with LSP and tree-sitter in mind. 

I've used GNU Emacs for 10 years with some thousands of lines of configuration. Lately I've been using neovim, where I have ~100 lines of config. I'm happy with helix with all of 5 lines of config, and LSP works better and faster with helix than any of my other editors. This team is on to something!

I tried editing a GNU Makefile today and found there is no syntax highlighting. This PR imports the preexisting `highlights.scm` file from [tree-sitter-make](https://github.com/alemuller/tree-sitter-make), as well as the tree-sitter-make submodule. I am also pretty new to tree-sitter so I didn't try my hand at writing the `indents.toml`, `injections.scm`, or the `textobjects.scm` files. Hopefully those are optional or we can iterate to flush out the functionality over time.

I'm not sure if you have any conventions for tracking existing `scm` files that are preferred over copy/paste, so just let me know if there are any changes requested. Cheers!

dead10ck commented on languages.toml:
```diff
@@ -387,6 +387,13 @@ comment-token = "#"
+[[language]]
+name = "make"
+scope = "source.make"
+file-types = ["Makefile"]
```
Could we also add the other variants, `makefile` and `GNUMakefile`? And also could we throw in `justfile` and `.justfile`?

EricCrosson replied:
Good thinking!

pickfire replied:
Wait, I thought there is tree-sitter-just? https://github.com/IndianBoy42/tree-sitter-just

dead10ck replied:
Oh interesting. Is it good to use? It mentions it's a work in progress and doesn't have tests yet.

dead10ck reviewed (commented):
Awesome, thank you!

archseer commented on runtime/queries/make/highlights.scm:
```diff
@@ -0,0 +1,172 @@
+
+(function_call
+  function: "warning"
+  (arguments (text) @text.warning))
```
`@warning`

archseer commented on runtime/queries/make/highlights.scm:
```diff
@@ -0,0 +1,172 @@
+
+(function_call
+  function: "info"
+  (arguments (text) @text.note))
```
`@info`

archseer reviewed (changes requested):
Thank you for the kind words!

The PR looks good but `highlights.scm` needs some changes to follow the list of scopes we use: https://docs.helix-editor.com/themes.html#syntax-highlighting

EricCrosson commented on runtime/queries/make/highlights.scm:
```diff
@@ -0,0 +1,172 @@
+
+(function_call
+  function: "error"
+  (arguments (text) @text.danger))
```
I don't see `@error` on [this list](https://docs.helix-editor.com/themes.html#syntax-highlighting) -- is this a scope that should also be added to `themes`.md?

EricCrosson commented:
> The PR looks good but `highlights.scm` needs some changes to follow the list of scopes we use: https://docs.helix-editor.com/themes.html#syntax-highlighting

Thanks for the pointer, and the complete review! I've updated the scopes and added `exception` to the themes documentation

Omnikar commented on languages.toml:
```diff
@@ -387,6 +387,13 @@ comment-token = "#"
+[[language]]
+name = "make"
+scope = "source.make"
+file-types = ["Makefile", "makefile", "justfile", ".justfile"]
```
`"justfile"` will cover files with `justfile` as their name or as their extension.
```suggestion
file-types = ["Makefile", "makefile", "justfile"]
```

dead10ck replied:
Does that include `.justfile` by itself?

Omnikar replied:
Ah, that is what this is meant to cover, my mistake.

pickfire replied:
I think we should remove just, there is https://github.com/IndianBoy42/tree-sitter-just

EricCrosson commented on .gitmodules:
```diff
@@ -177,7 +177,7 @@
 	path = helix-syntax/languages/tree-sitter-llvm-mir
 	url = https://github.com/Flakebi/tree-sitter-llvm-mir.git
-  shallow = true
+    shallow = true
```
Please note this addition, introduced intentionally while resolving a merge conflict

archseer commented:
Thanks! 🎉
